[Torch] Default to IPv6 on listen (#49124)

Babar opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/49124

At Facebook, most of our hosts are IPv6-only, but sometimes we do clowny things to make them dual-stacked.
As we do not expect them to be dual-stacked, they end up with an IPv4 IP address which is not in the DNS, and we do not have any `/etc/gai.conf` to order them.
So by default, if we try to bind anything to such a host, we'll get the IPv4 address.
But... if we try to connect to that host, the DNS will only ever resolve to IPv6, so we'll try to connect over v6, which won't work as the socket listens only on v4.

This will make the first try to use only IPv6. It should work in 99% of the cases, as long as the IPv6 stack is enabled.
Even if the host only has an IPv4 address, it will listen over v6 and use the compatibility layer (so `::ffff:` addresses)

And if the host has disabled IPv6 (using something like `ipv6.disable=1` on the kernel command line), we'll fallback to the old code, so it should be fully backwards compatible.

The main change is that IPv4 users might see "weird" addresses like `::ffff:127.0.0.1` instead of `127.0.0.1`, but otherwise, it should just work

Reviewed By: malfet

Differential Revision: D24969149



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

cbalioglu commented on test/distributed/test_c10d_gloo.py:
```diff
@@ -181,7 +181,7 @@ class RendezvousEnvTest(TestCase):
     def test_logging_init(self):
         os.environ["WORLD_SIZE"] = "1"
-        os.environ["MASTER_ADDR"] = "127.0.0.1"
+        os.environ["MASTER_ADDR"] = "::1"
```
This (and the following) test will likely fail on an IPv4-only machine. Another option would be to use `localhost` instead of an IP address.

Babar replied:
Yes, I've fixed those when I noticed the failures (switched to `localhost`)

cbalioglu commented on torch/csrc/distributed/c10d/Utils.cpp:
```diff
@@ -160,13 +171,18 @@ std::pair<int, PortType> listen(PortType port) {
   std::memset(&hints, 0x00, sizeof(hints));
   hints.ai_flags = AI_PASSIVE | AI_ADDRCONFIG;
-  hints.ai_family = AF_SELECTED; // IPv4 on Windows, IPv4/6 on Linux
+  hints.ai_family = AF_INET6; // default to IPv6
```
Another technique would be to specify `AF_UNSPEC` which requests `getaddinfo()` to return both IPv4 and IPv6 addresses and then do the filtering in the `while` loop on line 196 which I believe would be cleaner:

```
// Semi-pseudo code that prioritizes IPv6 over IPv4.
for (auto ai_family : [AF_INET6, AF_INET]) {
   while (nextAddr) {
        if (nextAddr->ai_family != ai_family) {
           continue;
        }
        ...
}
```

Babar replied:
That is similar to what specifying `/etc/gai.conf` would achieve, but without the side-effects.
That wouldn't work on Windows though, as it doesn't have `AF_UNSPEC`. This should work anywhere.

cbalioglu replied:
Another thing that I noticed after googling a bit is apparently on Windows the default for `IPV6_V6ONLY` is 1, meaning binding to a IPv6 socket on Windows (and if for some reason an admin changed it to 1 in sysctl on Linux) will disable IPv4 connections. I believe to keep BC we should also make sure that we call `setsockopt()` and explicitly set `IPV6_V6ONLY` to 0.

cbalioglu commented on torch/csrc/distributed/c10d/Utils.cpp:
```diff
@@ -278,18 +294,43 @@ int connect(
-  if (err != 0 || !res) {
+  // Append all the IPv4 addresses, we loop over them anyway
+  hints.ai_family = AF_INET;
+  int ipv4err = ::getaddrinfo(
```
Same comment as above.

Babar committed:
[Torch] Default to IPv6 on listen (#49124)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/49124

At Facebook, most of our hosts are IPv6-only, but sometimes we do clowny things to make them dual-stacked.
As we do not expect them to be dual-stacked, they end up with an IPv4 IP address which is not in the DNS, and we do not have any `/etc/gai.conf` to order them.
So by default, if we try to bind anything to such a host, we'll get the IPv4 address.
But... if we try to connect to that host, the DNS will only ever resolve to IPv6, so we'll try to connect over v6, which won't work as the socket listens only on v4.

This will make the first try to use only IPv6. It should work in 99% of the cases, as long as the IPv6 stack is enabled.
Even if the host only has an IPv4 address, it will listen over v6 and use the compatibility layer (so `::ffff:` addresses)

And if the host has disabled IPv6 (using something like `ipv6.disable=1` on the kernel command line), we'll fallback to the old code, so it should be fully backwards compatible.

The main change is that IPv4 users might see "weird" addresses like `::ffff:127.0.0.1` instead of `127.0.0.1`, but otherwise, it should just work

Reviewed By: malfet

Differential Revision: D24969149

fbshipit-source-id: ef42a07a0f921e2e4b08ef51c1ee797aa06caf94

cbalioglu commented:
See #68225.
